include/vapor/ArbitrarilyOrientedRegularGrid.h

```diff
@@ -0,0 +1,103 @@
+#pragma once
+
+#include <ostream>
+#include <vector>
+#include <vapor/common.h>
+#include <vapor/RegularGrid.h>
+#include <glm/glm.hpp>
+
+// clang-format off
+
+struct planeDescription {
+    size_t sideSize;
+    std::vector<double> origin;
+    std::vector<double> rotation;
+    VAPoR::CoordType boxMin;
+    VAPoR::CoordType boxMax;
+    VAPoR::CoordType domainMin;
+    VAPoR::CoordType domainMax;
+};
+
+namespace VAPoR {
+
+//! Create a 2D RegularGrid that is rotated along an arbitrary orientation.  This new grid
+//! samples a given 3D Grid object along a plane, through the entire domain.  The "planeDescription" 
+//! variable includes parameters that define the 2D plane's origin, rotation, and valid extents.  Points on the
+//! 2D RegularGrid that are outside of the valid extents will be assigned as
+//! missingValues.
+//!
+//! Note: The boxMin and boxMax values within the planeDescription should come from
+//! the current RenderParams' Box class.  The domainMin and domainMax values should
+//! come from the DataMgr::GetVariableExtents function.  Taking these values from
+//! the given 3D grid can produce erroneous results when the 3D grid's region is reduced.
+//!
+//! \param[in] grid3d A variable's 3D grid to be sampled along a plane, to generate a 2D grid object
+//! \param[in] description A set of std::vector<dobule> values that define the origin, rotation, and extents of the returned 2D grid
+//! \param[in] dims The sampling rate that the new grid will be defined upon, along its X and Y axes
+//! \param[out] data An array of sideSize*sideSize values that contain floating point values of the sampled data.
```

Comment:
Since we've take the trouble to create this class, I think we should NOT pass in a `data` argument, but instead have the class manage the space internally. Ideally, all of the Grid classes would not require a `blks` argument, but this has significant implications for DataMgr caching that do not apply here. 

Yes, this would mean that the SliceRenderer would no longer have a back door into the classes internal memory, but it probably shouldn't anyhow.

Reply:
I thought about this too.  But unless I'm missing something, the only way to pass a Grid a blks pointer is through its constructor.  

ArbitrarilyOrientedRegularGrid's constructor receives its blks pointer and passes it to the RegularGrid constructor via the initialization list.  Without a function like Grid::SetBlks(), is it possible for ArbitrarilyOrientedRegularGrid's constructor to allocate data and configure it as is done in its parent classes?

Reply:
Unless I'm mistaken you can call `new` within the constructor member initializer

Reply:
That makes sense.  This last issue is in progress.  It's at the top of my list and should be submitted tomorrow morning barring anything unforeseen.

Reply:
@clyne - I'm finding problems implementing this and I'm not sure it can be done without some kind of Grid::SetBlks() function.  

The fundamental thing we're trying to do is to allocate memory (blks) in ArbitrarilyOrientedRegularGrid's initialization list.

I've tried using raw and smart pointers to allocate the memory within the RegularGrid's constructor argument list.  

**Problem with smart pointers:** A smart pointer that's created in an argument list will delete the allocated data as soon as it falls out of scope, which is at the end of the RegularGrid's constructor.

```
ArbitrarilyOrientedRegularGrid::ArbitrarilyOrientedRegularGrid(
    const VAPoR::Grid *grid3d,
    planeDescription& pd,
    const DimsType& dims
) : RegularGrid(
        dims,
        {{pd.sideSize, pd.sideSize, 1}},
        {(std::unique_ptr<float>(new float[dims[0]*dims[1]*dims[2]])).get()},    // memory gets deleted after initialization list finishes
        {{0.,0.,0.}},
        {{1.,1.,1.}}
) {...}
```

**Problem with raw pointers:** Allocating memory with a raw pointer as a constructor argument will assign no lvalue that can be deleted later, so we get a memory leak.

```
ArbitrarilyOrientedRegularGrid::ArbitrarilyOrientedRegularGrid(
    const VAPoR::Grid *grid3d,
    planeDescription& pd,
    const DimsType& dims
) : RegularGrid(
        dims,
        {{pd.sideSize, pd.sideSize, 1}},
        {new float[dims[0]*dims[1]*dims[2]]},    // How do we delete this in the destructor?
        {{0.,0.,0.}},
        {{1.,1.,1.}}
) {...}
```

I've tried using private smart pointers in ArbitrarilyOrientedRegularGrid so they don't lose scope until the destructor is called (in the case of smart pointers).  This does not work because I can't allocate their memory before the RegularGrid constructor is called, which is where we need the memory.  Private raw pointers that we can manually delete will not work for the same reason - we can't allocate them before RegularGrid's constructor.

I feel like I'm running out of ideas aside from implementing a Grid::SetBlks() method, or Grid::initialize().

Reply:
Take a look at `PDoubleInput.cpp +6`

Reply:
@stasj - thanks, this does work!

Followup question: I cannot get this to work with smart pointers without std::make_unique, introduced in c++14. 
 Is this the reason PDoubleInput uses raw pointers instead of smart pointers?